go/libraries/doltcore/table/untyped/csv/reader.go

```diff
@@ -153,32 +204,200 @@ func (csvr *CSVReader) Close(ctx context.Context) error {
+// The result is only valid until the next call to readLine.
+func (csvr *CSVReader) readLine() ([]byte, error) {
+	line, err := csvr.bRd.ReadSlice('\n')
+	if err == bufio.ErrBufferFull {
+		csvr.rawBuffer = append(csvr.rawBuffer[:0], line...)
+		for err == bufio.ErrBufferFull {
+			line, err = csvr.bRd.ReadSlice('\n')
+			csvr.rawBuffer = append(csvr.rawBuffer, line...)
+		}
+		line = csvr.rawBuffer
+	}
+	if len(line) > 0 && err == io.EOF {
+		err = nil
+		// For backwards compatibility, drop trailing \r before EOF.
+		if line[len(line)-1] == '\r' {
+			line = line[:len(line)-1]
+		}
+	}
+	csvr.numLine++
+	// Normalize \r\n to \n on all input lines.
+	if n := len(line); n >= 2 && line[n-2] == '\r' && line[n-1] == '\n' {
+		line[n-2] = '\n'
+		line = line[:n-1]
 	}
+	return line, err
+}
 
-	taggedVals := make(row.TaggedValues)
-	for i := 0; i < allCols.Size(); i++ {
-		col := allCols.GetByIndex(i)
-		if colVals[i] == nil {
-			taggedVals[col.Tag] = nil
-			continue
+func (csvr *CSVReader) csvReadRecords(dst []*string) ([]*string, error) {
+	var keepString []bool
+
+	// Read line (automatically skipping past empty lines and any comments).
+	var line, fullLine []byte
+	var errRead error
+	for errRead == nil {
+		line, errRead = csvr.readLine()
+		if errRead == nil && len(line) == lengthNL(line) {
+			line = nil
+			continue // Skip empty lines
 		}
-		taggedVals[col.Tag] = types.String(*colVals[i])
+		fullLine = line
+		break
+	}
+	if errRead == io.EOF {
+		return nil, errRead
+	}
+
+	// Parse each field in the record.
+	var err error
+	const quoteLen = len(`"`)
+	delimLen := len(csvr.delim)
+	recLine := csvr.numLine // Starting line for record
+	csvr.recordBuffer = csvr.recordBuffer[:0]
+	csvr.fieldIndexes = csvr.fieldIndexes[:0]
+parseField:
+	for {
+		line = bytes.TrimLeftFunc(line, unicode.IsSpace)
+		if len(line) == 0 || line[0] != '"' {
+			// Non-quoted string field
+			i := bytes.Index(line, csvr.delim)
+			field := line
+			if i >= 0 {
+				field = field[:i]
+			} else {
+				field = field[:len(field)-lengthNL(field)]
+			}
+			csvr.recordBuffer = append(csvr.recordBuffer, field...)
+			csvr.fieldIndexes = append(csvr.fieldIndexes, len(csvr.recordBuffer))
+			// discard unquoted empty strings
+			keepString = append(keepString, len(field) != 0)
+			if i >= 0 {
+				line = line[i+delimLen:]
+				continue parseField
+			}
+			break parseField
+		} else {
+			// Quoted string field
+			line = line[quoteLen:]
+			for {
+				i := bytes.IndexByte(line, '"')
+				if i >= 0 {
+					// Hit next quote.
+					csvr.recordBuffer = append(csvr.recordBuffer, line[:i]...)
+					line = line[i+quoteLen:]
+					if atDelim(line, csvr.delim) {
+						// `"<delimiter>` sequence (end of field).
+						line = line[delimLen:]
+						csvr.fieldIndexes = append(csvr.fieldIndexes, len(csvr.recordBuffer))
+						keepString = append(keepString, true)
+						continue parseField
+					} else {
+						switch rn := nextRune(line); {
```

Comment:
I'm not sure about this switch statement.  Switches should either be switched on something:

    switch x {
        case 1:
        case 2:
        ...
    }

or should be switched without an expression

    switch {
        case x == 1:
        case x == 2:
        ...
    }

now you can switch on something you initialize in the switch:

    switch x := f(); x {
        case 1:
        case 2:
    }

and I already think that starts to look confusing conflating the initialization with the value you are switching on, but using the semicolon to initialize a value and then switching on an empty expression in order to save 1 line of code isn't worth the confusion.

    rn := nextRune(line)
    switch {
        ...

is better.